SoundWire: add support for ACE2.x

plbossart opened the pull request:
This PR uses the abstraction added in past kernel cycles to provide support for the ACE2.x integration. The existing SHIM and Cadence registers are now split in 3 (SHIM, IP, SHIM vendor-specific), with some parts also moved to the HDaudio Extended Multi link structures. Nothing fundamentally different except for the register map.

This PR only provides the basic mechanisms to expose SoundWire-based DAIs. The PCI parts and DSP management will be contributed later, and the DAI ops are now empty as well.

Note the dependencies between ASoC and SoundWire, this will need specific attention when upstreaming.

ranj063 commented on drivers/soundwire/intel_shim2.c:
```diff
@@ -155,8 +156,52 @@ static void intel_debugfs_init(struct sdw_intel *sdw) {}
+
+	mutex_lock(sdw->link_res->shim_lock);
+
+	sdw->cdns.link_up = false;
```
is this on purpose that we want to mark the link are powered down even if the power_down call below fails?

plbossart replied:
I must admit I don't know or don't recall.

It's the same pattern as in the existing intel.c so I just followed without thinking,

The only place where this value is used is in the interrupt thread.

````c
	/* Check if the link is up */
	if (!cdns->link_up)
		return IRQ_NONE;
````

I think it makes sense to use this variable as a sign that you cannot deal with an interrupt while you are powering down.

It still sounds rather racy to me.

@bardliao do you remember anything about this?

bardliao replied:
@plbossart @ranj063 IIRC, we set `sdw->cdns.link_up = false;` before actual power down and don't set it back even if fail to power down on purpose. 
See the comment below.
```
                        /*
                         * we leave the sdw->cdns.link_up flag as false since we've disabled
                         * the link at this point and cannot handle interrupts any longer.
                         */
```
That is to avoid superfluous interrupt handling.

plbossart replied:
@ranj063 ok with no change?

ujfalusi replied:
probably add  the same comment here?

ranj063 commented on drivers/soundwire/intel_shim2.c:
```diff
@@ -158,17 +158,52 @@ static void intel_debugfs_exit(struct sdw_intel *sdw) {}
+	struct sdw_bus *bus = &sdw->cdns.bus;
+	u32 *shim_mask = sdw->link_res->shim_mask;
+	struct sdw_master_prop *prop = &bus->prop;
+	u32 syncprd;
```
nit-pick: can we follow our coding style of structs first and then the others?

plbossart replied:
sure.

plbossart replied:
done

ranj063 commented on drivers/soundwire/intel_shim2.c:
```diff
@@ -158,17 +158,52 @@ static void intel_debugfs_exit(struct sdw_intel *sdw) {}
+	if (!*shim_mask) {
+		/* we first need to program the SyncPRD/CPU registers */
+		dev_dbg(sdw->cdns.dev,
+			"first link up, programming SYNCPRD\n");
```
one line?

plbossart replied:
yes

plbossart replied:
done

ranj063 commented on drivers/soundwire/intel_shim2.c:
```diff
@@ -490,6 +490,17 @@ static int intel_register_dai(struct sdw_intel *sdw)
+	int ret;
+
+	ret = hdac_bus_eml_sdw_set_lsdiid(sdw->link_res->hbus, sdw->instance, dev_num);
+
```
newline unnecessary

plbossart replied:
ok

plbossart replied:
done

ranj063 reviewed (commented):
nice work @plbossart. I have a few minor comments and one about whether thte BSD-3 license applies to the file with debugfs parts

plbossart commented:
pushed v2 based on @ranj063 feedback, also moved debugfs code to separate file with GPL2.0-only license. Also moved inclusion of all headers earlier

[v2.diff.txt](https://github.com/thesofproject/linux/files/10762408/v2.diff.txt)

RanderWang commented on include/linux/soundwire/sdw_intel.h:
```diff
@@ -101,6 +105,77 @@
+#define SDW_SHIM2_PCMSCAP_BSS		GENMASK(12, 8)	/* Bidirectional streams */
+
+/* Read-only PCM Stream Channel Count, y variable is stream */
+#define SDW_SHIM2_PCMSYCHC(y)		(0x14 + (0x4 * (y)))
```
0x14 or 0x16 ?  I see "#define SDW_SHIM2_PCMSYCHM(y)		(0x16 + (0x4 * (y)))"

plbossart replied:
yes it's fine @RanderWang 

the read-only channel count is at offset 0x14
the channel-map is at offset 0x16, see line 134

The last letter matters, it's CHC (count) v. CHM (Map).

RanderWang commented on drivers/soundwire/intel_shim2.c:
```diff
@@ -0,0 +1,24 @@
+#include <linux/soundwire/sdw_intel.h>
+#include "cadence_master.h"
+#include "bus.h"
+#include "intel.h"
```
Are all these header files used in this file now ?  or in future ?

plbossart replied:
These were initially in a wrong place, I moved them here to avoid multiple updates. I could add them gradually if this was desired, but I don't see how that would help.

RanderWang commented on drivers/soundwire/intel_shim2_debugfs.c:
```diff
@@ -0,0 +1,147 @@
+{
+	int value;
+
+	if (l)
```
what is l ?  I first considered it as if(1).

plbossart replied:
l -> long. this tells you if the code will read a u32 or u16 register. this is the same code as in the existing intel.c, so I'd prefer to leave it as is.

RanderWang commented on drivers/soundwire/intel_shim2.c:
```diff
@@ -115,10 +115,174 @@ static int intel_link_power_down(struct sdw_intel *sdw)
+	int i;
+
+	for (i = 0; i < num_pdi; i++) {
+		pdi->ch_count = intel_pdi_get_ch_cap(sdw, pdi->num);
```
do we need to check ch_count is valid or not ?   If we don't need to check it, why not change the function qualifier to void ? we don't check the return value of this function in intel_pdi_stream_ch_update

plbossart replied:
good suggestion, I can change the return type of intel_pdi_get_ch_update() and intel_pdi_stream_ch_update().

plbossart replied:
done

plbossart commented:
Adressed @RanderWang in v3 : 
[v3.diff.txt](https://github.com/thesofproject/linux/files/10768471/v3.diff.txt)

RanderWang reviewed (approved):
LGTM

plbossart commented:
v4: one-line conflict with upstream, fixed now.

plbossart commented:
CI results show a failure in kmod load/unload tests, tentatively fixed by https://github.com/thesofproject/sof-test/pull/1004

plbossart commented:
Now that https://github.com/thesofproject/sof-test/pull/1004 is merged, the tests should pass.

plbossart commented:
SOFCI TEST

plbossart commented:
Same issue with  "This branch cannot be rebased due to too many changes"

That's really bad, as if any PR above 15 commits is just too much work?

ujfalusi commented on drivers/soundwire/Makefile:
```diff
@@ -20,7 +20,7 @@ soundwire-cadence-y := cadence_master.o
 
 #Intel driver
-soundwire-intel-y :=	intel.o intel_auxdevice.o intel_init.o dmi-quirks.o \
+soundwire-intel-y :=	intel.o  intel_shim2.o intel_auxdevice.o intel_init.o dmi-quirks.o \
```
it is strange to see `intel_shim2.o` without shim1, if I'm not wrong the intel.o is the intel_shim1?
Just observation.

plbossart replied:
intel_ace2x would be much better indeed

plbossart replied:
done

ujfalusi commented on sound/soc/sof/intel/hda.c:
```diff
@@ -175,6 +175,17 @@ static int hda_sdw_probe(struct snd_sof_dev *sdev)
+						  AZX_REG_ML_LEPTR_ID_SDW);
+		if (!eml_lock)
+			return -ENODEV;
+		res.eml_lock = eml_lock;
```
nitpick: There is no need for the temporary `eml_lock`
```c
		res.eml_lock = hdac_bus_eml_get_mutex(sof_to_bus(sdev), true,
						      AZX_REG_ML_LEPTR_ID_SDW);
		if (!res.eml_lock)
			return -ENODEV;
```

plbossart replied:
agree

plbossart replied:
done

ujfalusi commented on drivers/soundwire/intel_shim2_debugfs.c:
```diff
@@ -0,0 +1,147 @@
+#include <linux/pm_runtime.h>
+#include <linux/soundwire/sdw_registers.h>
+#include <linux/soundwire/sdw.h>
+#include <linux/soundwire/sdw_intel.h>
```
nitpick^2: these could be in alphabetic order.

plbossart replied:
agree

plbossart replied:
done

ujfalusi commented on drivers/soundwire/intel_shim2_debugfs.c:
```diff
@@ -0,0 +1,147 @@
+	ssize_t ret;
+	int j;
+	u32 pcm_cap;
+	int pcm_bd;
```
nitpick: ordering of these to form a reverse christmas tree?

plbossart replied:
agree

plbossart replied:
done

ujfalusi commented on drivers/soundwire/intel_shim2.c:
```diff
@@ -10,12 +10,57 @@
 #include <linux/soundwire/sdw_registers.h>
 #include <linux/soundwire/sdw.h>
 #include <linux/soundwire/sdw_intel.h>
+#include <sound/hda-mlink.h>
```
Since you already introducing bunch of includes when adding the file, why not add this in there as well? Or add them one by one as needed.

plbossart replied:
I added what's needed for each file, and here this is the first time hda-mlink.c is actually needed.

ujfalusi commented on drivers/soundwire/intel_shim2.c:
```diff
@@ -17,17 +17,51 @@
+		}
+	}
+
+	*shim_mask |= BIT(link_id);
```
Why can't we use `sdw->instance` or should `hdac_bus_eml_sdw_power_up_unlocked()` be also using `link_id`?

plbossart replied:
yes, it's inconsistent, will fix

plbossart replied:
done

ujfalusi commented on drivers/soundwire/intel_shim2.c:
```diff
@@ -110,6 +110,42 @@ static int intel_link_power_down(struct sdw_intel *sdw)
+
+	ret = hdac_bus_eml_sdw_sync_go_unlocked(sdw->link_res->hbus);
+
+	if (ret < 0)
```
nitpick: no need for the blank line

plbossart replied:
ok

plbossart replied:
done

ujfalusi commented on drivers/soundwire/intel_shim2.c:
```diff
@@ -31,6 +31,41 @@ static void intel_shim_vs_init(struct sdw_intel *sdw)
+{
+	void __iomem *shim_vs = sdw->link_res->shim_vs;
+	u16 wake_en;
+	u16 wake_sts;
```
nitpick: either swap them or move to single line? (the `wake_sts` is only used if `!wake_enable`)

plbossart replied:
I'd rather keep that order since the enable and sts are listed in that order in the register map.

ujfalusi reviewed (commented):
@plbossart, I can only contribute (if I recall correctly) with some nitpicks, the series is really nice to read.

ujfalusi commented:
> Same issue with "This branch cannot be rebased due to too many changes"
> 
> That's really bad, as if any PR above 15 commits is just too much work?

Should we create a GH ticket for this? I don't think this low patch number limitation is in any way sane.

plbossart commented:
v4: addressed @ujfalusi 's comments, except for the two not marked as resolved.

[v4.diff.txt](https://github.com/thesofproject/linux/files/10817818/v4.diff.txt)

plbossart commented:
> > Same issue with "This branch cannot be rebased due to too many changes"
> > That's really bad, as if any PR above 15 commits is just too much work?
> 
> Should we create a GH ticket for this? I don't think this low patch number limitation is in any way sane.

I think so, if this continues it's a problem for us really.

plbossart commented:
One more round of approvals needed @ujfalusi @ranj063 @bardliao @RanderWang. The last version should be good-enough for merge now.

ujfalusi commented on drivers/soundwire/intel_ace2x_debugfs.c:
```diff
@@ -0,0 +1,147 @@
+DEFINE_DEBUGFS_ATTRIBUTE(intel_set_s_datamode_fops, NULL,
+			 intel_set_s_datamode, "%llu\n");
+
+void intel_shim2_debugfs_init(struct sdw_intel *sdw)
```
We are using ace2.x and shim2 interchangeably? ACE2.x comes with a new SHIM layer, shim2, right?

No need to change anything, just to clarify for me.

plbossart replied:
no, this is confusing indeed, because we have the SHIM and SHIM_VS (vendor-specific) areas. using shim2 is just confusing. I'll use ace2x instead

plbossart replied:
done

ujfalusi commented on drivers/soundwire/intel_ace2x.c:
```diff
@@ -10,12 +10,57 @@
+
+	ret = hdac_bus_eml_sdw_power_down_unlocked(sdw->link_res->hbus, sdw->instance);
+	if (ret < 0) {
+		dev_err(sdw->cdns.dev, "%s: hdac_bus_eml_sdw_power_down failed: %d\n",
```
nitpick: lift the comment @bardliao mentioned?
https://github.com/thesofproject/linux/blob/398a19cef5d447537438bebc69579d89d8ec7ebd/drivers/soundwire/intel.c#L456-L459

plbossart replied:
make sense

plbossart replied:
done

ujfalusi commented on drivers/soundwire/intel_ace2x.c:
```diff
@@ -15,6 +15,22 @@
+ * shim_vs ops
+ */
+
+static void intel_shim_vs_init(struct sdw_intel *sdw)
```
First I read this as "shim **VS** init" ;)
What is 'shim_vs'?

plbossart replied:
shim vendor-specific. I'll clarify, if it's confusing then someone else will also have the same issue.

plbossart replied:
done

ujfalusi reviewed (approved):
@plbossart, only one ask to lift a comment in, the other comments does not need any action (but would be nice to get answer to them).

Looks nice.

RanderWang reviewed (dismissed):
LGTM

plbossart commented:
v5 changes: 
[v5.diff.txt](https://github.com/thesofproject/linux/files/10843147/v5.diff.txt)

plbossart commented:
going for a merge commit to work-around the "This branch cannot be rebased due to too many changes" issue
